ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/worldstate/MarkSweepPrunerTest.java

```diff
@@ -45,26 +58,109 @@ public void shouldMarkAllNodesInCurrentWorldState() {
+                worldStateStorage,
+                new WorldStatePreimageKeyValueStorage(new InMemoryKeyValueStorage()))
+            .getMutable();
+    final MutableBlockchain blockchain = mock(DefaultBlockchain.class);
+    final BlockHeader header = mock(BlockHeader.class);
+    final MarkSweepPruner pruner =
+        new MarkSweepPruner(
+            worldStateStorage, blockchain, new InMemoryKeyValueStorage(), metricsSystem);
+
+    // Generate accounts and save corresponding state root
+    gen.createRandomContractAccountsWithNonEmptyStorage(worldState, 20);
+    final Hash stateRoot = worldState.rootHash();
+
+    when(blockchain.getBlockHeader(0)).thenReturn(Optional.of(header));
+    when(header.getStateRoot()).thenReturn(stateRoot);
+
+    // Nothing is marked so we should sweep everything, but we need to make sure the state root goes
+    // first
+    pruner.sweepBefore(1);
+    InOrder inOrder = inOrder(hashValueStore, stateStorage);
+    inOrder.verify(hashValueStore).remove(stateRoot);
+    inOrder.verify(stateStorage).removeUnless(any());
+  }
+
+  @Test
+  public void shouldntRemoveStateRootIfItsMarked() {
+    // Setup "remote" state
+    final InMemoryKeyValueStorage markStorage = new InMemoryKeyValueStorage();
+    final InMemoryKeyValueStorage stateStorage = new InMemoryKeyValueStorage();
+    final WorldStateStorage worldStateStorage = new WorldStateKeyValueStorage(stateStorage);
+    final WorldStateArchive worldStateArchive =
+        new WorldStateArchive(
+            worldStateStorage,
+            new WorldStatePreimageKeyValueStorage(new InMemoryKeyValueStorage()));
+    final MutableWorldState worldState = worldStateArchive.getMutable();
+    final MutableBlockchain blockchain = mock(DefaultBlockchain.class);
+    final BlockHeader header = mock(BlockHeader.class);
+    final MarkSweepPruner pruner =
+        new MarkSweepPruner(worldStateStorage, blockchain, markStorage, metricsSystem);
+
+    // Generate accounts and save corresponding state root
+    gen.createRandomContractAccountsWithNonEmptyStorage(worldState, 20);
+    final Hash stateRoot = worldState.rootHash();
+
+    when(blockchain.getBlockHeader(1)).thenReturn(Optional.of(header));
+    when(header.getStateRoot()).thenReturn(stateRoot);
+
+    gen.createRandomContractAccountsWithNonEmptyStorage(worldState, 10);
+    final BlockHeader preSweepHeader = mock(BlockHeader.class);
+
+    when(blockchain.getBlockHeader(0)).thenReturn(Optional.of(preSweepHeader));
+    final Hash preSweepStateRoot = worldState.rootHash();
+    when(preSweepHeader.getStateRoot()).thenReturn(preSweepStateRoot);
+
+    pruner.mark(stateRoot);
+    pruner.mark(preSweepStateRoot);
+    pruner.flushPendingMarks();
+    final Set<BytesValue> markedNodes = new HashSet<>(markStorage.keySet());
+
+    // All those new nodes should be removed when we sweep
+    pruner.sweepBefore(1);
+    assertThat(stateStorage.keySet()).containsExactlyInAnyOrderElementsOf(markedNodes);
```

Comment:
This is a little hard to follow - can you add explicit assertions on the state roots you don't expect to be removed: `assertThat(stateStorage.containsKey(preSweepStateRoot));`

Reply:
Good point, as it stands this test's scope overlaps with another test's scope right now anyway so it can be narrowed.